tests/security/389ds/tls_389ds_server.pm

```diff
@@ -0,0 +1,112 @@
+# Copyright (C) 2021 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+#
+# Summary: Implement & Integrate 389ds + sssd test case into openQA,
+#          This test module covers the server setup processes
+#
+# Maintainer: rfan1 <richard.fan@suse.com>
+# Tags: poo#88513, tc#1768672
+
+use base 'consoletest';
+use testapi;
+use strict;
+use warnings;
+use utils;
+use lockapi;
+use mmapi 'wait_for_children';
+
+sub run {
+    select_console("root-console");
+
+    my $local_ip    = '10.0.2.101';
+    my $remote_ip   = '10.0.2.102';
+    my $local_name  = '389ds';
+    my $remote_name = 'sssdclient';
+    my $ca_dir      = '/etc/openldap/ssl';
+    my $inst_ca_dir = '/etc/dirsrv/slapd-localhost';
+
+    # Install 389-ds and create an server instance
+    zypper_call("in 389-ds");
+    assert_script_run("wget --quiet " . data_url("389ds/instance.inf") . " -O /tmp/instance.inf");
+    assert_script_run("dscreate from-file /tmp/instance.inf");
+    validate_script_output("dsctl localhost status", sub { m/Instance.*is running/ });
+
+    # Configure CA Certificates for TLS
+    assert_script_run("wget --quiet " . data_url("389ds/tls_cert.sh") . " -O /tmp/tls_cert.sh");
+    assert_script_run("wget --quiet " . data_url("389ds/.dsrc") . " -O /root/.dsrc");
+    assert_script_run("bash /tmp/tls_cert.sh $local_name $ca_dir");
+    assert_script_run("certutil -D -d $inst_ca_dir -n Server-Cert");
+    assert_script_run("certutil -D -d $inst_ca_dir -n Self-Signed-CA");
+    assert_script_run("dsctl localhost tls import-server-key-cert $ca_dir/server.pem $ca_dir/server.key");
+    assert_script_run("dsctl localhost tls import-ca $ca_dir/myca.pem myca");
+    assert_script_run("cp $ca_dir/myca.pem $inst_ca_dir/ca.crt");
+    systemctl("restart dirsrv\@localhost.service");
+
+    # Configure host names for C/S communication
+    assert_script_run("sed -i '/$local_ip.*/d' /etc/hosts");
+    assert_script_run("sed -i '/$remote_ip.*/d' /etc/hosts");
+    assert_script_run("echo '$local_ip $local_name.example.com $local_name' >> /etc/hosts");
+    assert_script_run("echo '$remote_ip $remote_name.example.com $remote_name' >> /etc/hosts");
+    assert_script_run("hostname $local_name");
+
+    # Create ldap user and group,generate sample sssd.conf file
+    my $ldap_user    = 'wilber';
+    my $ldap_passwd  = 'pw_389ds';
```

Comment:
We have already `$testapi::username` and `$testapi::password`.

Reply:
Nice comments, let me have a try

Reply:
Cool, it worked fine, thanks much